Gems/PhysX/Code/Source/JointComponent.cpp

```diff
@@ -90,28 +90,110 @@ namespace PhysX
         {
-            if (m_configuration.m_followerEntity == m_configuration.m_leadEntity)
+            AZ_Error(
+                "JointComponent::Activate()",
+                false,
+                "Joint's lead entity cannot be the same as the entity in which the joint resides. Joint failed to initialize.");
+            return;
+        }
+
+        // If joint has no lead entity, it is a constraint on a global frame (position & orientation).
+        const bool hasLeadEntity = m_configuration.m_leadEntity.IsValid();
+
+        // Collect the involved entities.
+        m_rigidBodyEntityMap.insert({ m_configuration.m_followerEntity, false });
+        if (hasLeadEntity)
+        {
+            m_rigidBodyEntityMap.insert({ m_configuration.m_leadEntity, false });
+        }
+
+        // Connect to RigidBodyNotificationBus of follower and leader rigid bodies
+        // and wait until all of them are enabled to create the native joint.
+        Physics::RigidBodyNotificationBus::MultiHandler::BusConnect(m_configuration.m_followerEntity);
+        if (hasLeadEntity)
+        {
+            Physics::RigidBodyNotificationBus::MultiHandler::BusConnect(m_configuration.m_leadEntity);
+        }
+    }
+
+    void JointComponent::Deactivate()
+    {
+        if (m_rigidBodyEntityMap.empty())
+        {
+            return;
+        }
+
+        DestroyNativeJoint();
+
+        Physics::RigidBodyNotificationBus::MultiHandler::BusDisconnect();
+        m_rigidBodyEntityMap.clear();
+    }
+
+    void JointComponent::OnPhysicsEnabled(const AZ::EntityId& entityId)
+    {
+        auto it = m_rigidBodyEntityMap.find(entityId);
+        if (it != m_rigidBodyEntityMap.end())
+        {
+            it->second = true;
+
+            const bool allRigidBodiesEnabled = AZStd::all_of(
+                m_rigidBodyEntityMap.begin(),
+                m_rigidBodyEntityMap.end(),
+                [](const auto& elem)
+                {
+                    return elem.second;
+                });
+
+            if (allRigidBodiesEnabled)
             {
-                AZ_Error("JointComponent::Activate()",
-                    false,
-                    "Joint's lead entity cannot be the same as the entity in which the joint resides. Joint failed to initialize.");
-                return;
+                CreateNativeJoint();
             }
+        }
+    }
+
+    void JointComponent::OnPhysicsDisabled(const AZ::EntityId& entityId)
+    {
+        auto it = m_rigidBodyEntityMap.find(entityId);
+        if (it != m_rigidBodyEntityMap.end())
+        {
+            it->second = false;
 
-            AZ::EntityBus::Handler::BusConnect(m_configuration.m_followerEntity);
+            DestroyNativeJoint();
         }
     }
 
-    void JointComponent::Deactivate()
+    void JointComponent::CreateNativeJoint()
     {
-        AZ::EntityBus::Handler::BusDisconnect();
+        if (m_jointHandle != AzPhysics::InvalidJointHandle)
+        {
+            return;
```

Comment:
Should this be an assertable condition? i.e. is it expected that this call is idempotent?

Reply:
In this case it's important that these `CreateNativeJoint/DestroyNativeJoint` functions validate if it's already created/destroyed the joint before doing anything because they come from physics bus calling enable/disable on the 1 or 2 entities that it relies on.

Technically speaking the `CreateNativeJoint` wouldn't need it, but the `DestroyNativeJoint` yes. But it's much better when both functions are symmetric and rely on the same logic (easier to understand the code), plus these functions are private, so they serve the purpose of internal logic, no need to answer to external users calling them.